### DISCUSSION TITLE
Bluetooth management

### DISCUSSION THREAD

**Discussion by alessandro-satanassi** [TeamOpenFIRE/OpenFIRE-Firmware#19] [General]
I found the reason why with the bluetooth initialization an error is reported when trying to insert the three profiles (keyboard, absmouse and gamepade16) in the 'description report' (desc_bt_report) ... the problem is not in the Tiny_USBdevices.cpp code, but in the arduino-pico library in the PICOBluetoothHID.h file at line 247 (uint8_t _hid_service_buffer[300];). The buffer is too small and the long definition string of the 3 devices does not fit. Just increase the size of the buffer and the problem is solved. Changing it with uint8_t _hid_service_buffer[1300]; .. the initialization occurs correctly and the three devices are recognized by the operating system. (I think even fewer bytes are enough). In the meantime I continue with the study and arrangement of the part regarding the connection via bluetooth.

**Comment by alessandro-satanassi** [comment]
Maybe I'm wrong, but I think I discovered that with the 'arduino-pico' libraries it is not possible to use 'serialBT' and 'PICOBluetoothHID' at the same time, unless you change part of the code of these libraries. Can you confirm this? Do you know if there is any other way or do you have any other ideas? ('PICOBluetoothHID' is needed to create the keyboard, mouse and gampad devices, while 'serialBT' is needed to make the App communicate with the gun and also for MameHooker' during the game

**Reply by SeongGino** [reply]
Isn't that your job to investigate with the code that you volunteered to look into? I don't know why you're asking me.

**Reply by alessandro-satanassi** [reply]
You're right, but I thought someone had already looked at it. I'll take care of it.

**Comment by alessandro-satanassi** [comment]
Could I try to create a Bluetooth library specifically for OpenFire by merging and partially modifying the code of the two libraries arduino-pico serialBT and PICObluetoothHID? (I thought the bluetooth support in the arduino-pico libraries was more advanced)

**Reply by SeongGino** [reply]
Only if you're okay being the one maintaining the library with future iterations of `arduino-pico`; messing with internal libraries can be a problem if the implem ever changes in upstream.

We already have to maintain a fork with a one-liner difference just to work around a bug in TinyUSB wrt serial UART, so I'd rather not exacerbate the issue if possible.

**Reply by alessandro-satanassi** [reply]
I didn't think of forking arduino-pico, but implementing a library to keep in the OpenFire project, like TinyUSB_Devices, where I reimplement the management of the PICOBloutoothHID and serialBT libraries, using more or less the same code of that library (the primitive routines would remain in arduino-pico and I wouldn't touch them). I'll try and then let you know how it goes. Of course I would commit to maintaining that library.
.. by the way, a curiosity if you can answer me quickly: why is your fork of arduino-pico not from the main library of 'earlephilhower', but from another fork of 'anrp-tri'. Are there other changes made by him too?

**Comment by alessandro-satanassi** [comment]
As promised, I'm implementing Bluetooth management, both for the HID part (keyboard, mouse, gampad emulation ... which was basically already done, except for a problem with the gamepad due to problems with the arduino-pico libraries), and for the serial connection via Bluetooth for 'Mamehooker' and communication with the OpenFire App on the PC. I'm still working on it and I had to do a lot of testing, because there is little documentation on Bluetooth management on the Pico W. I have little time and so I'm going slowly.
I'm trying to implement Bluetooth management by making as few changes as possible to the code.
I wanted to ask 'SeongGino', as the creator and maintainer of the code, something so I can also orient myself based on his directives/preferences.
The information messages sent to the terminal with Serial.print, Serial.println, etc. (example: Serial.println("Exec run mode"), do you prefer them to be sent also via Bluetooth when the gun is connected only via Bluetooth or do you prefer them to be sent only via USB serial when the gun is connected with the USB cable?

**Reply by SeongGino** [reply]
I'm not sure what this is asking; any messages that are sent should be being sent through whichever connection method is being used.

**Comment by SeongGino** [comment]
With regards to lockups when Bluetooth mode is enabled, it might not actually be (fully) related to the BTstack at all; see https://github.com/TeamOpenFIRE/OpenFIRE-Firmware/issues/22
